metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java

```diff
@@ -607,4 +612,143 @@ public void testEarlyControllerResults() throws Throwable {
+                List<CreatableReplicaAssignment> partitions = IntStream
+                    .range(0, numPartitions)
+                    .mapToObj(partitionIndex -> new CreatableReplicaAssignment()
+                        .setPartitionIndex(partitionIndex)
+                        .setBrokerIds(Arrays.asList(0, 1, 2))
+                    )
+                    .collect(Collectors.toList());
+
+                Uuid topicId = controller.createTopics(
+                    new CreateTopicsRequestData()
+                        .setTopics(
+                            new CreatableTopicCollection(
+                                Collections.singleton(
+                                    new CreatableTopic()
+                                        .setName(topicName)
+                                        .setNumPartitions(-1)
+                                        .setReplicationFactor((short) -1)
+                                        .setAssignments(new CreatableReplicaAssignmentCollection(partitions.iterator()))
+                                ).iterator()
+                            )
+                        )
+                ).get().topics().find(topicName).topicId();
+
+                // Create a lot of alter isr
+                List<AlterIsrRequestData.PartitionData> alterIsrs = IntStream
+                    .range(0, numPartitions)
+                    .mapToObj(partitionIndex -> {
+                        PartitionRegistration partitionRegistration = controller.replicationControl().getPartition(
+                            topicId,
+                            partitionIndex
+                        );
+
+                        return new AlterIsrRequestData.PartitionData()
+                            .setPartitionIndex(partitionIndex)
+                            .setLeaderEpoch(partitionRegistration.leaderEpoch)
+                            .setCurrentIsrVersion(partitionRegistration.partitionEpoch)
+                            .setNewIsr(Arrays.asList(0, 1));
+                    })
+                    .collect(Collectors.toList());
+
+                AlterIsrRequestData.TopicData topicData = new AlterIsrRequestData.TopicData()
+                        .setName(topicName);
+                topicData.partitions().addAll(alterIsrs);
+
+                int leaderId = 0;
+                AlterIsrRequestData alterIsrRequest = new AlterIsrRequestData()
+                    .setBrokerId(leaderId)
+                    .setBrokerEpoch(brokerEpochs.get(leaderId));
+                alterIsrRequest.topics().add(topicData);
+
+                logEnv.logManagers().get(0).resignAfterNonAtomicCommit();
+
+                assertThrows(
+                    ExecutionException.class,
+                    () -> controller.alterIsr(alterIsrRequest).get()
+                );
+
+                // Wait for the new active controller
+                final QuorumController newController = controlEnv.activeController();
```

Comment:
This confused me a little bit since we are trying to verify that the state on the original controller resets properly. That is what is happening here since there is only one controller in the test, but it is obscured a little bit by the new variable. Maybe it would be clearer to use the original reference and write this as:
```java
assertEquals(controller, controlEnv.activeController());
```
Also, is there an epoch or something we can bump to ensure the transition?

Reply:
Good idea. Fixed.